### PR TITLE
feat: Typeorm modification (MAPCO-4071 , MAPCO-4089 , MAPCO-4090)

### PR DESCRIPTION
**Pull request by RonitKissis**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✔                                                                       |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                      |
| Chore            | ✖                                                                       |

Further  information:

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,32 +1,28 @@
...
-import { ConnectionManager } from './DAL/connectionManager';
 
 const port: number = config.get<number>('server.port') || DEFAULT_SERVER_PORT;
 
-const app = getApp();
+void getApp()
+  .then(([app, container]) => {
+    const logger = container.resolve<Logger>(SERVICES.LOGGER);
+    const healthCheck = container.resolve<HealthCheck>(SERVICES.HEALTH_CHECK);
+    // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (netanelC) [inline src/index.ts:18]: Unnecessary

File: src/index.ts [hunk 2]
```diff
@@ -1,32 +1,28 @@
-// this import must be called before the first import of tsyring
+/* eslint-disable import/first */
```
Comment (netanelC) [inline src/index.ts:1]: Why did you add this?

File: tests/integration/heartbeat/heartbeat.spec.ts [hunk 3]
```diff
@@ -1,89 +1,80 @@
...
     });
 
     it('removeHeartbeats should return 200 status code and remove records from db', async () => {
       const ids = ['id1', 'id2'];
-      const expectedIds = ids.map((id) => ({ id }));
-      heartbeatRepositoryMocks.removeMock.mockResolvedValue(expectedIds);
-
       const response = await requestSender.removeHeartbeats(ids);
 
+      //expect(response).toSatisfyApiSpec();
```
Comment (netanelC) [inline tests/integration/heartbeat/heartbeat.spec.ts:76]: Remove this comment

File: tests/integration/heartbeat/heartbeat.spec.ts [hunk 4]
```diff
@@ -1,89 +1,80 @@
...
+    await depContainer.resolve(DataSource).destroy();
   });
 
   describe('Happy Path', function () {
+    it('pulse should return 200 status code and save the heartbeat pulse', async function () {
+      const id = '1';
+      saveSpy = jest.spyOn(repo, 'save');
+      const response = await requestSender.pulse(id);
+
+      //expect(response).toSatisfyApiSpec();
```
Comment (netanelC) [inline tests/integration/heartbeat/heartbeat.spec.ts:51]: Remove this comment

File: tests/configurations/integration/jest.globalSetup.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { initConnection } from '../../../src/DAL/utils/createConnection';
+
+export default async (): Promise<void> => {
+  const dataSourceOptions = config.get<IDbConfig>('typeOrm');
+  const connection = await initConnection({
+    ...dataSourceOptions,
+    migrationsTableName: 'custom_migration_table',
+    migrations: ['src/DAL/migrations/*.ts'],
+  });
+  const schema = `"${dataSourceOptions.schema}"` || undefined;
```
Comment (netanelC) [inline tests/configurations/integration/jest.globalSetup.ts:12]: The schema will never be undefined

File: tests/configurations/integration/jest.globalTeardown.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import config from 'config';
+import { IDbConfig } from '../../../src/common/interfaces';
+import { initConnection } from '../../../src/DAL/utils/createConnection';
+
+export default async (): Promise<void> => {
+  const dataSourceOptions = config.get<IDbConfig>('typeOrm');
+  const connection = await initConnection(dataSourceOptions);
+  const schema = `"${dataSourceOptions.schema}"` || undefined;
```
Comment (netanelC) [inline tests/configurations/integration/jest.globalTeardown.ts:8]: The schema will never be undefined

File: README.md [hunk 7]
```diff
@@ -77,3 +77,15 @@ To only run integration tests:
...
+## Running Migration Docker
+
+To run the migration docker run the following command:
+
+```bash
+
+ docker run -it -e DB_USER=postgres  -e DB_PASSWORD=1234 -e DB_NAME=postgres -e DB_TYPE=postgres -e DB_HOST=localhost -e DB_PORT=5432 -e DB_SCHEMA=HeartbeatManager --rm --network host migration-image:vtag
+
+```
+
```
Comment (netanelC) [inline README.md:90]: Unnecessary empty line

File: README.md [hunk 8]
```diff
@@ -77,3 +77,15 @@ To only run integration tests:
...
+
```
Comment (netanelC) [inline README.md:91]: Unnecessary empty line

File: dataSource.migrations.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const { DataSource } = require('typeorm');
+const config = require('config');
+const { createConnectionOptions } = require('./src/DAL/utils/createConnection');
+
+/**
+ * @type {import("./src/db/types").DbConfig}
+ */
+const connectionOptions = config.get('typeOrm');
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
```
Comment (netanelC) [inline dataSource.migrations.js:10]: Is this necessary?

File: dataSource.migrations.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+const appDataSource = new DataSource({
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (netanelC) [inline dataSource.migrations.js:12]: Is this necessary?

**Review by netanelC — APPROVED** [review]
An impressive PR, keep up the hard work! most of the comments are lint issues so I won't block the PR but handle them before you're merging to the master branch

File: tests/configurations/integration/jest.globalSetup.ts (hunk 5 above)
Comment (netanelC) [reply tests/configurations/integration/jest.globalSetup.ts:12]:
```suggestion
  const schema = dataSourceOptions.schema !== undefined ? `"${dataSourceOptions.schema}"` : 'public';
```

File: README.md (hunk 8 above)
Comment (RonitKissis) [reply README.md:91]: thank you mister linter

File: src/DAL/utils/createConnection.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  return {
+    entities: [HeartbeatEntity],
+    migrationsTableName: 'custom_migration_table',
+    ...dataSourceOptions,
+    type: 'postgres',
+  };
+};
+
+export const initConnection = async (dbConfig: IDbConfig): Promise<DataSource> => {
+  const dataSource = new DataSource(createConnectionOptions(dbConfig));
```
Comment (asafMasa) [inline src/DAL/utils/createConnection.ts:26]: is it possible to add log here?

File: src/index.ts [hunk 12]
```diff
@@ -1,32 +1,28 @@
...
 import config from 'config';
 import { DEFAULT_SERVER_PORT, SERVICES } from './common/constants';
+
 import { getApp } from './app';
-import { ConnectionManager } from './DAL/connectionManager';
 
 const port: number = config.get<number>('server.port') || DEFAULT_SERVER_PORT;
 
-const app = getApp();
+void getApp()
```
Comment (asafMasa) [inline src/index.ts:14]: is it possible to add log here of - "starting app"

File: tests/integration/heartbeat/heartbeat.spec.ts (hunk 4 above)
Comment (asafMasa) [reply tests/integration/heartbeat/heartbeat.spec.ts:51]: revert remark keep code

File: tests/integration/heartbeat/heartbeat.spec.ts [hunk 13]
```diff
@@ -1,89 +1,80 @@
...
-        id: id,
-        lastHeartbeat: nowDate,
-      } as unknown as HeartbeatEntity;
-
-      const response = await requestSender.pulse(id);
-
-      expect(response.status).toBe(httpStatusCodes.OK);
-      expect(heartbeatRepositoryMocks.saveMock).toHaveBeenCalledTimes(1);
-      expect(heartbeatRepositoryMocks.saveMock).toHaveBeenCalledWith(entity);
+      expect(findSpy).toHaveBeenCalledTimes(1);
```
Comment (asafMasa) [inline tests/integration/heartbeat/heartbeat.spec.ts:69]: add expect(response).toSatisfyApiSpec();

File: tests/integration/heartbeat/heartbeat.spec.ts (hunk 3 above)
Comment (asafMasa) [reply tests/integration/heartbeat/heartbeat.spec.ts:76]: revert remark keep code

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 1 above)
Comment (RonitKissis) [reply src/index.ts:18]: wrong, eslint throws error on /livenss

File: src/index.ts (hunk 12 above)
Comment (RonitKissis) [reply src/index.ts:14]: what about line 20?

File: dataSource.migrations.js (hunk 10 above)
Comment (RonitKissis) [reply dataSource.migrations.js:12]: removed

File: src/DAL/repositories/heartbeatRepository.ts [hunk 14]
```diff
@@ -1,41 +1,38 @@
...
 
-@EntityRepository(HeartbeatEntity)
-export class HeartbeatRepository extends Repository<HeartbeatEntity> {
-  private readonly appLogger: Logger; //don't override internal repository logger.
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+const createHeartbeatRepository = (dataSource: DataSource) => {
+  return dataSource.getRepository(HeartbeatEntity).extend({
+    async pulse(id: string): Promise<void> {
+      const entity = this.create({ id: id, lastHeartbeat: new Date() });
+      await this.save(entity);
```
[the remark this answers is not in the document]
Comment (RonitKissis) [reply src/DAL/repositories/heartbeatRepository.ts:11]: cant log here but this class is called by the manager and there are logs there

File: src/DAL/repositories/heartbeatRepository.ts [hunk 15]
```diff
@@ -1,41 +1,38 @@
...
-  public async pulse(id: string): Promise<void> {
-    const entity = new HeartbeatEntity({
-      id: id,
-      lastHeartbeat: new Date(),
-    });
-    await this.save(entity);
-  }
+    async removeHeartbeats(ids: string[]): Promise<number> {
+      const entities = ids.map((id) => ({ id })) as unknown as HeartbeatEntity[];
+      const deleted = await this.remove(entities);
```
[the remark this answers is not in the document]
Comment (RonitKissis) [reply src/DAL/repositories/heartbeatRepository.ts:26]: cant log here but this class is called by the manager and there are logs there

File: src/DAL/repositories/heartbeatRepository.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (RonitKissis) [reply src/DAL/repositories/heartbeatRepository.ts:11]: log exists in heartbeatManager

File: src/DAL/repositories/heartbeatRepository.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (RonitKissis) [reply src/DAL/repositories/heartbeatRepository.ts:26]: log exists in heartbeatManager

File: src/DAL/utils/createConnection.ts (hunk 11 above)
Comment (RonitKissis) [reply src/DAL/utils/createConnection.ts:30]: added

File: tests/integration/heartbeat/heartbeat.spec.ts (hunk 4 above)
Comment (RonitKissis) [reply tests/integration/heartbeat/heartbeat.spec.ts:51]: added a to do due to the bug we discussed

File: tests/integration/heartbeat/heartbeat.spec.ts (hunk 13 above)
Comment (RonitKissis) [reply tests/integration/heartbeat/heartbeat.spec.ts:69]: added a to do due to the bug we discussed

File: tests/integration/heartbeat/heartbeat.spec.ts (hunk 3 above)
Comment (RonitKissis) [reply tests/integration/heartbeat/heartbeat.spec.ts:77]: added a to do due to the bug we discussed

File: src/common/utils.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+export class TimeoutError extends Error {}
```
[the remark this answers is not in the document]
Comment (RonitKissis) [reply src/common/utils.ts:1]: can you explain the purpose of that? Heartbeat manager doesnt use mc-utils

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good Job!

**Review by asafMasa — APPROVED** [review]

**Pull request merged**